### ISSUE TITLE
Incorrect result with the default eqsolver settings

### ISSUE THREAD

**Issue by TheGreatfpmK** [stormchecker/storm#1076]
This is a bit of a wild one, and it might not necessarily be a bug in Storm itself.

For model and props: [model.zip](https://github.com/user-attachments/files/32599013/model.zip) 

the default settings of Storm return value: 7.548904625e+180, though it should be basically close to 0.

# Claude analysis:

## Storm bug: default solver silently returns a wildly wrong reachability reward on a DTMC

`model.prism` is a DTMC with 183,620 states and 673,273 transitions.
`model.props` checks `R{"lostH"}=? [F "finished"]`. `P=? [F "finished"]` is 1, so the expected
reward is finite.

- **Expected:** `--exact` gives **≈ 2.97314993e-21**, i.e. essentially 0.
- **Actual, with Storm's default settings:** **7.548904625e+180**, about 201 orders of magnitude
  off. The only symptom is this warning:
  `WARN (GmmxxLinearEquationSolver.cpp:134): Iterative solver did not converge within 22 iteration(s).`

Found with Storm 1.14.0 (dev), commit `6e32f9cb598c10a8b0d5306171f2563f51bce3c8` (2026-09-17).

## Isolating the trigger

The defaults are `--eqsolver topological`, `--topological:eqsolver gmm++`, `--gmm++:method gmres`
and `--gmm++:precond ilu`. Each row changes one thing from those defaults:

| Options (added to the defaults)  | Result                                                    |
|----------------------------------|-----------------------------------------------------------|
| *(none)*                         | **7.548904625e+180** (wrong)                              |
| `--gmm++:precond none`           | 2.973151526e-21                                           |
| `--gmm++:precond diagonal`       | 2.973151526e-21                                           |
| `--gmm++:method bicgstab`        | 1.244646011e-23                                           |
| `--gmm++:method qmr`             | crashes: `QMR failed to converge` (gmm_solver_qmr.h:156)  |
| `--gmm++:restart 1000`           | **7.548904625e+180** (wrong), still stops at 22 iterations |
| `--gmm++:maxiter 100000`         | **7.548904625e+180** (wrong), still stops at 22 iterations |
| `--topological:eqsolver native`  | 9.794222458e-13                                           |
| `--eqsolver gmm++` (no topological) | 2.525876637e-08                                        |
| `--eqsolver native`              | 1.147471541e-16                                           |
| `--eqsolver eigen`               | 0                                                         |
| `--exact`                        | ≈ 2.97314993e-21 (exact rational result)                  |

What this suggests:

- **The ILU preconditioner is the trigger.** With `none` or `diagonal`, GMRES gets the right answer.
- **It isn't an iteration limit.** GMRES stops at iteration 22, below both the restart threshold
  (default 50) and the iteration cap, and raising either one changes nothing. It looks like a
  breakdown that gets reported only as "did not converge". The non-converged vector is then
  used as the result anyway. QMR with the same preconditioner throws an exception instead, so it
  at least fails loudly.
- **It is only catastrophic inside the topological solver.** Plain gmm++ on the whole system
  gives 2.5e-08. Presumably one SCC's subsystem breaks ILU+GMRES, and its garbage values are
  propagated to the SCCs that depend on it.
- **The gmm++ options do take effect** under topological, even though Storm warns
  `gmm++ is not selected as the preferred equation solver, so setting options for gmm++ might
  have no effect.` That warning is misleading here.

## Reproduce

Each of these runs in about a second:

```
storm --prism model.prism --prop model.props                          # 7.548904625e+180 (wrong)
storm --prism model.prism --prop model.props --gmm++:precond none     # 2.973151526e-21
storm --prism model.prism --prop model.props --exact                  # ~2.97314993e-21 (ground truth)
```

Or with stormpy's default environment:

```python
import stormpy
prog = stormpy.parse_prism_program("model.prism")
props = stormpy.parse_properties_for_prism_program('R{"lostH"}=? [F "finished"]', prog)
model = stormpy.build_model(prog, props)
result = stormpy.model_checking(model, props[0])
print(result.at(model.initial_states[0]))  # 7.548904625090014e+180, should be ~2.97e-21
```

## How it was found

PAYNT's SMPMC synthesis engine checks many assignments of this sketch. It trusted this value as a
genuine violation of `R{"lostH"} <= 0.01` for an assignment that actually satisfies it.

**Comment by TheGreatfpmK** [comment]
Prop: `R{"lostH"}=? [ F "finished" ]`

PRISM program:
```
dtmc

// To make it more challenging, we can extend the model and consider low and  high priority requests
// using two request queues.

const int CMAX = 100;

const int X11 = 0;
const int X21 = 0;
const int X31 = 0;
const int X41 = 0;
const int X12 = 0;
const int X22 = 0;
const int X32 = 0;
const int X42 = 0;
const int X13 = 0;
const int X23 = 0;
const int X33 = 0;
const int X43 = 0;
const int X14 = 0;
const int X24 = 0;
const int X34 = 0;
const int X44 = 0;


const int LT1 = 3;
const int LT2 = 5;
const int LT3 = 7;
const int HT1 = 3;
const int HT2 = 5;
const int HT3 = 7;

// This system can achieve

//R{"power"}<= 18000 [ F (bat=0) ]

//R{"lostH"}<= 0.3 [ F (bat=0) ]

//R{"lostL"}<= 0.6 [ F (bat=0) ]


module CLOCK

	c : [0..1];

	[tick1] bat>0 & c=0 -> (c'=1);
	[tick2] bat>0 & c=1 -> (c'=0);

endmodule

// POWER MANAGER

module PM

	pm  :  [0..3] init 3;
	// 0 - go to active
	// 1 - go to active for LOW
	// 2 - go to idle
	// 3 - go to sleep


	[tick1] qL*10 <= LT1*CMAX & q*10 <= HT1*CMAX   -> (pm'=X11);

	[tick1] qL*10 > LT1*CMAX & qL*10 <= LT2*CMAX &  q*10 <= HT1*CMAX -> (pm'=X21);

	[tick1] qL*10 > LT2*CMAX & qL*10 <= LT3*CMAX &  q*10 <= HT1*CMAX -> (pm'=X31);

    [tick1] qL*10 > LT3*CMAX &  q*10 <= HT1*CMAX -> (pm'=X41);

    [tick1] qL*10 <= LT1*CMAX & q*10 > HT1*CMAX & q*10 <= HT2*CMAX -> (pm'=X12);

    [tick1] qL*10 > LT1*CMAX & qL*10 <= LT2*CMAX &  q*10 > HT1*CMAX & q*10 <= HT2*CMAX -> (pm'=X22);

	[tick1] qL*10 > LT2*CMAX & qL*10 <= LT3*CMAX &  q*10 > HT1*CMAX & q*10 <= HT2*CMAX -> (pm'=X32);

    [tick1] qL*10 > LT3*CMAX &  q*10 > HT1*CMAX & q*10 <= HT2*CMAX -> (pm'=X42);

    [tick1] qL*10 <= LT1*CMAX & q*10 > HT2*CMAX & q*10 <= HT3*CMAX -> (pm'=X13);

	[tick1] qL*10 > LT1*CMAX & qL*10 <= LT2*CMAX &  q*10 > HT2*CMAX & q*10 <= HT3*CMAX -> (pm'=X23);

	[tick1] qL*10 > LT2*CMAX & qL*10 <= LT3*CMAX &  q*10 > HT2*CMAX & q*10 <= HT3*CMAX -> (pm'=X33);

    [tick1] qL*10 > LT3*CMAX &  q*10 > HT2*CMAX & q*10 <= HT3*CMAX -> (pm'=X43);

    [tick1] qL*10 <= LT1*CMAX & q*10 > HT3*CMAX   -> (pm'=X14);

	[tick1] qL*10 > LT1*CMAX & qL*10 <= LT2*CMAX &  q*10 > HT3*CMAX -> (pm'=X24);

	[tick1] qL*10 > LT2*CMAX & qL*10 <= LT3*CMAX &  q*10 > HT3*CMAX -> (pm'=X34);

    [tick1] qL*10 > LT3*CMAX &  q*10 > HT3*CMAX -> (pm'=X44);

endmodule

// SERVICE PROVIDER

module SP

	sp : [0..8] init 3;
	// 0 active
	// 1 active for LOW
	// 2 idle
	// 3 sleep

	// 4 active_both_to_sleep
	// 5 idle_to_sleep

	// 6 sleep_to_idle
	// 7 sleep_to_active
	// 8 sleep_to_active_L

	// states where PM has no control (transient states)

	[tick2] sp=4  -> 0.9  : (sp'=4) + 0.1  : (sp'=3);
	[tick2] sp=5  -> 0.9   : (sp'=5) + 0.1   : (sp'=3);

	[tick2] sp=6  -> 0.9   : (sp'=6) + 0.1   : (sp'=2);

	[tick2] sp=7  -> 0.9  : (sp'=7) + 0.1  : (sp'=0);
	[tick2] sp=8  -> 0.9  : (sp'=8) + 0.1  : (sp'=1);


	// states where PM has control
	// goto_active
	[tick2] sp=0 & pm=0 -> (sp'=0); // from active
	[tick2] sp=1 & pm=0 -> (sp'=0); // from activeL
	[tick2] sp=2 & pm=0 -> (sp'=0); // from idle
	[tick2] sp=3 & pm=0 -> (sp'=7); // from sleep


	// goto_active_L
	[tick2] sp=0 & pm=1 -> (sp'=1); // from active
	[tick2] sp=1 & pm=1 -> (sp'=1); // from activeL
	[tick2] sp=2 & pm=1 -> (sp'=1); // from idle
	[tick2] sp=3 & pm=1 -> (sp'=8); // from sleep

	// goto_idle
	[tick2] sp=0 & pm=2 -> (sp'=2); // from active
	[tick2] sp=1 & pm=2 -> (sp'=2); // from active_L
	[tick2] sp=2 & pm=2 -> (sp'=2); // from idle
	[tick2] sp=3 & pm=2 -> (sp'=6); // from sleep

	// goto_sleep
	[tick2] sp=0 & pm=3 -> (sp'=4); // from active
	[tick2] sp=1 & pm=3 -> (sp'=4); // from active_L
	[tick2] sp=2 & pm=3 -> (sp'=5); // idle
	[tick2] sp=3 & pm=3 -> (sp'=3); // sleep

endmodule


// SERVICE REQUESTER
module SR

	//a two-queue varinat

	sr : [0..2] init 0; // 0 - idle, 1 - Hreq, 2-Lreq

	[tick2] sr=0 -> 0.898: (sr'=0) + 0.051: (sr'=1) + 0.051: (sr'=2);
	[tick2] sr=1 -> 0.254: (sr'=0) + 0.373: (sr'=1) + 0.373: (sr'=2);
	[tick2] sr=2 -> 0.254: (sr'=0) + 0.373: (sr'=1) + 0.373: (sr'=2);

endmodule


// SERVICE REQUEST QUEUES
module SRQ

	q : [0..CMAX] init 0;
	qL : [0..CMAX] init 0;

	[tick2] sr=0 & sp!=0 & sp!=1 -> true; // do not serve and nothing arrives

	[tick2] sr=1 & sp!=0 & sp!=1 -> (q'=min(q+1,CMAX)); // do not serve and a request arrives
	[tick2] sr=2 & sp!=0 & sp!=1 -> (qL'=min(qL+1,CMAX)); // do not serve and a request arrives

	[tick2] sr=0 & sp=0 -> (q'=max(q-1,0)); // serve and nothing arrives
	[tick2] sr=0 & sp=1 -> (qL'=max(qL-1,0)); // serve and nothing arrives

	// serve and a request arrives arrives -- it actually performs two actions
	[tick2] sr=2 & sp=0 -> (q'=max(q-1,0)) & (qL'=min(qL+1,CMAX));
	[tick2] sr=1 & sp=1 -> (qL'=max(qL-1,0)) & (q'=min(q+1,CMAX));

	[tick2] sr=1 & sp=0 -> true;
	[tick2] sr=2 & sp=1 -> true;

endmodule





// BATTERY
module BAT

	bat : [0..1] init 1; // 0 empty, 1 - operational

	[] bat=0 -> (bat'=0); // loop when battery empty

	[tick2] bat=1 -> 0.01 : (bat'=0) + 0.99 : (bat'=1);

endmodule

// queue size
rewards "queueH"

	c=1 : q;

endrewards

rewards "queueL"

	c=1 : qL;

endrewards

// customers lost
rewards "lostH"

	[tick2] q=CMAX & sr=1 & sp!=0 : 1;

endrewards

rewards "lostL"

	[tick2] qL=CMAX & sr=2 & sp!=1 : 1;

endrewards


// power consumption
rewards "power"

// in contrasr to the PRISM model we assume that the transient states have the power consuption
// equal to the source state

	sp=1 & c=1 : 2.5;
	sp=0 & c=1 : 2.5;
	sp=4 & c=1 : 2.5;

	sp=2 & c=1 : 1.5;
	sp=5 & c=1 : 1.5;

	sp=3 & c=1 : 0.1;
	sp=6 & c=1 : 0.1;
	sp=7 & c=1 : 0.1;
	sp=8 & c=1 : 0.1;

endrewards

// time
rewards "time"

    [tick2] true : 1;

endrewards

label "finished" = bat = 0;
```
